beps/00010-event-auditor/README.md

```diff
@@ -0,0 +1,221 @@
+
+To ensure consistency and facilitate regulatory compliance, the proposal suggests creating a shared package that defines a data model for audit events. This model consists of several key components.
+
+#### Actor Details Interface
+
+This interface defines the information related to the actor who triggered the logged event. It includes fields like actor ID, IP address, hostname, and user agent.
+
+```ts
+export type ActorDetails = {
+  actorId: string | null;
+  ip?: string;
+  hostname?: string;
+  userAgent?: string;
+};
+```
+
+#### Audit Request/Response Interfaces
+
+These interfaces define the structure of request and response data that might be included in the audit event. It's important to note that these interfaces exclude sensitive information like tokens from headers or other irrelevant details to avoid security risks.
+
+```ts
+export type AuditRequest = {
+  body: any;
+  url: string;
+  method: string;
+  params?: any;
+  query?: any;
+};
+
+export type AuditResponse = {
+  status: number;
+  body?: any;
+};
+```
+
+#### Audit Event Status Interfaces
+
+These interfaces define the possible statuses for an audit event entry. There are three options:
+
+```ts
+/**
+ * Indicates the event was successful.
+ */
+export type AuditEventSuccessStatus = { status: 'succeeded' };
+
+/**
+ * Indicates the event failed and includes details about the encountered errors.
+ */
+export type AuditEventFailureStatus = {
+  status: 'failed';
+  errors: ErrorLike[];
+};
+
+/**
+ * Indicates the event failed with errors that don't perfectly fit the expected structure.
+ */
+export type AuditEventUnknownFailureStatus = {
+  status: 'failed';
+  errors: unknown[];
+};
+
+export type AuditEventStatus = AuditEventSuccessStatus | AuditEventFailureStatus;
+```
+
+#### EventAuditor Interface
+
+This interface defines the functionalities of an `EventAuditor` class. This class provides methods for:
+- Extracting the actor ID from an Express request (if available).
+- Creating detailed audit event information based on provided options.
+- Logging an audit event with a specific level (info, debug, warn, or error).
+
+```ts
+/**
+ * Common fields of an audit event. Note: timestamp and pluginId are automatically added at event creation.
+ *
+ * @public
+ */
+export type AuditEventDetails = {
+  actor: ActorDetails;
+  eventName: string;
+  stage: string;
+  request?: AuditRequest;
+  response?: AuditResponse;
+  meta: JsonValue;
+  isAuditEvent: true;
+} & AuditEventStatus;
+
+export type AuditEventDetailsOptions = {
+  eventName: string;
+  stage: string;
+  metadata?: JsonValue;
+  response?: AuditResponse;
+  actorId?: string;
+  request?: Request;
+} & (AuditEventSuccessStatus | AuditEventUnknownFailureStatus);
+
+export type AuditEventOptions = {
+  eventName: string;
+  message: string;
+  stage: string;
+  level?: 'info' | 'debug' | 'warn' | 'error';
+  actorId?: string;
+  metadata?: JsonValue;
+  response?: AuditResponse;
+  request?: Request;
```

Comment:
This is a little bit tricky since it may encourage request objects to be passed pretty deep down into plugin code. Perhaps it's best to keep things oriented around metadata instead and e.g. resource IDs?

Reply:
The main reason for the `request` is to be able to extract information like `ip`, `hostname`, backstage identity, and endpoint url (and request details).

Reply:
Yep let's settle https://github.com/backstage/backstage/pull/25060#discussion_r1670971604 before any other discussion imo, as that really sets the direction for where the touch points are for this facility